File: fixtures/typescript/dynamic-eval/negative.ts
// Scores a model, queries a database, and schedules work. None evaluates a
// string.

function score(model, batch) {
  // Must NOT fire: a mode switch, not an evaluator.
  model.eval();
  return model(batch);
}

function query(cursor, sql) {
  // Must NOT fire: a cursor method; also how every regex in this language runs.
  return cursor.exec(sql);
}

function later(fn) {
  // Must NOT fire: the function form evaluates nothing, and it is the form
  // nearly every asynchronous skill uses. Matching the callee name alone would
  // report all of them.
  return setTimeout(fn, 100);
}

function parse(text) {
  // Must NOT fire: deserialization is not evaluation.
  return JSON.parse(text);
}

module.exports = { score, query, later, parse };

File: fixtures/python/dynamic-eval/negative.py
"""Scores a model and parses configuration. Neither turns a string into code."""

import json


def score(model, batch):
    # Must NOT fire: PyTorch's mode switch. This means "stop training", and it
    # appears across a large fraction of ML-adjacent skills — the single false
    # positive that would make this rule unshippable.
    model.eval()
    return model(batch)


def query(cursor, sql):
    # Must NOT fire: a database cursor method.
    return cursor.exec(sql)


def load(text):
    # Must NOT fire: deserialization is not evaluation.
    return json.loads(text)
